sql/catalyst/src/main/java/org/apache/spark/sql/connector/read/PartitionReader.java

```diff
@@ -51,7 +51,8 @@
   T get();
 
   /**
-   * Returns an array of custom task metrics. By default it returns empty array.
+   * Returns an array of custom task metrics. By default it returns empty array. Note that it is
+   * not recommended to put heavy logic in this method as it may affect reading performance.
```

Comment:
Normally, this method should be very cheap, as it just returns the current metrics values that are being tracked. The implementation may track the metrics per-row or per-batch, but that's out of our control.

That said, even if we call this method only at the end of partition read, the perf overhead can still be big if the data source tracks the metrics per-row under the hood.

@wypoon does it answer your concern?

Reply:
My interest is in the batch scan use case. It is true that, under the hood, a particular DS v2 may well need to track some particular metric per row, but there are definitely cases where a metric does not need to be updated per row; in those cases, simply reporting the metric value at the end of the read would be cheaper.

Reply:
I think it is worth spelling out when `PartitionReader#currentMetricsValues()` is called on the DS v2, and how the values are used (in this case, they are used to set the long in the `SQLMetric`, not added to it).

Reply:
Since we can't control how frequently the underlying data source tracks the metrics, this is all about how frequently we want to update the SQL metrics.

Since the actual metrics update is sent to the driver via heartbeat, it's not very useful if we update the metrics per-row. How about we do it like per 100 rows? Only update once may be problematic if the task runs for a long time.

Reply:
If the iterator is not consumed to the end, doesn't it mean we will produce inaccurate metrics if it stops in the middle of 100?

Reply:
That's a good point. For whole stage codegen with limit, we may stop earlier before consuming the entire iterator.

If we want to do per 100 rows update, we also need to do one more update at the end of the task.

Reply:
It sounds like the DS v2 should be prepared to compute and provide the metrics whenever `PartitionReader#currentMetricsValues()` is called. I think if we always call it at the end of the task (not precluding calling it more often), we should be able to get accurate metrics.

Reply:
So sounds like a feasible approach is to call `currentMetricsValues` and update metrics per some rows and also update it at the end of the task?

Reply:
yea I think so

Reply:
We can probably add the "update metrics per some rows" in a followup as it's an improvement.

Reply:
okay, sgtm.